django/db/migrations/graph.py

```diff
@@ -79,6 +79,29 @@ def descendants(self):
         return self.__dict__['_descendants']
 
 
+class DummyNode(Node):
+    def __init__(self, key, origin, error_message):
+        super(DummyNode, self).__init__(key)
+        self.origin = origin
+        self.error_message = error_message
+
+    def __repr__(self):
+        return '<DummyNode: (%r, %r)>' % self.key
+
+    def promote(self):
+        """
+        Transition dummy to a normal node and clean off excess attribs.
+        Creating a Node object from scratch would be too much of a
+        hassle as many dependendies would need to be remapped.
+        """
+        del self.origin
+        del self.error_message
+        self.__class__ = Node
```

Comment:
This pattern is actually used in multiple places in Django:

```
$ ack --py -C 3 "\.__class__ = "
django/contrib/gis/gdal/field.py
33-        self.ptr = fld_ptr
34-
35-        # Setting the class depending upon the OGR Field Type (OFT)
36:        self.__class__ = OGRFieldTypes[self.type]
37-
38-        # OFTReal with no precision should be an OFTInteger.
39-        if isinstance(self, OFTReal) and self.precision == 0:
40:            self.__class__ = OFTInteger
41-            self._double = True
42-
43-    def __str__(self):

django/contrib/gis/gdal/geometries.py
117-            self.srs = srs
118-
119-        # Setting the class depending upon the OGR Geometry Type
120:        self.__class__ = GEO_CLASSES[self.geom_type.num]
121-
122-    def __del__(self):
123-        "Deletes this Geometry."

django/contrib/gis/geos/geometry.py
112-                6: MultiPolygon,
113-                7: GeometryCollection,
114-            }
115:        self.__class__ = GEOSGeometry._GEOS_CLASSES[self.geom_typeid]
116-
117-        # Setting the coordinate sequence for the geometry (will be None on
118-        # geometries that do not have coordinate sequences)

django/db/models/fields/__init__.py
86-
87-def _empty(of_cls):
88-    new = Empty()
89:    new.__class__ = of_cls
90-    return new
91-
92-
--
487-        # We need to avoid hitting __reduce__, so define this
488-        # slightly weird copy construct.
489-        obj = Empty()
490:        obj.__class__ = self.__class__
491-        obj.__dict__ = self.__dict__.copy()
492-        return obj
493-

django/db/models/sql/query.py
261-        used by clients to update attributes after copying has taken place.
262-        """
263-        obj = Empty()
264:        obj.__class__ = klass or self.__class__
265-        obj.model = self.model
266-        obj.alias_refcount = self.alias_refcount.copy()
267-        obj.alias_map = self.alias_map.copy()

django/utils/tree.py
40-        any extra setting up to do).
41-        """
42-        obj = Node(children, connector, negated)
43:        obj.__class__ = cls
44-        return obj
45-
46-    def __str__(self):
--
55-        Utility method used by copy.deepcopy().
56-        """
57-        obj = Node(connector=self.connector, negated=self.negated)
58:        obj.__class__ = self.__class__
59-        obj.children = copy.deepcopy(self.children, memodict)
60-        return obj
61-

tests/basic/tests.py
376-                return 'did something'
377-
378-        qs = Article.objects.all()
379:        qs.__class__ = CustomQuerySet
380-        qs = qs.none()
381-        with self.assertNumQueries(0):
382-            self.assertEqual(len(qs), 0)
--
653-            def get_queryset(self):
654-                return FakeQuerySet(self.model)
655-        try:
656:            Article._base_manager.__class__ = FakeManager
657-            asos = ArticleSelectOnSave.objects.create(pub_date=datetime.now())
658-            with self.assertNumQueries(3):
659-                asos.save()
--
666-            with self.assertRaises(DatabaseError):
667-                asos.save(update_fields=['pub_date'])
668-        finally:
669:            Article._base_manager.__class__ = orig_class
670-
671-
672-class ModelRefreshTests(TestCase):
```